Source/WTF/wtf/WeakPtr.h

```diff
@@ -200,49 +205,69 @@ template<typename T, typename Counter = EmptyCounter> class WeakPtrFactory {
 
     ~WeakPtrFactory()
     {
-        if (!m_impl)
-            return;
-        m_impl->clear();
+        if (auto* pointer = m_impl.pointer())
+            pointer->clear();
+    }
+
+    CounterType* counter()
+    {
+        if (auto* pointer = m_impl.pointer())
+            return &pointer->counter();
+        return nullptr;
+    }
+
+    CounterType* counter() const
+    {
+        if (auto* pointer = m_impl.pointer())
+            return &pointer->counter();
+        return nullptr;
     }
 
     void initializeIfNeeded(const T& object) const
     {
-        if (m_impl)
+        if (m_impl.pointer())
             return;
 
         ASSERT(m_wasConstructedOnMainThread == isMainThread());
-        m_impl = WeakPtrImpl<Counter>::create(const_cast<T*>(&object));
+        m_impl.setPointer(WeakPtrImpl<CounterType>::create(const_cast<T*>(&object)));
     }
 
     template<typename U> WeakPtr<U, Counter> createWeakPtr(U& object, EnableWeakPtrThreadingAssertions enableWeakPtrThreadingAssertions = EnableWeakPtrThreadingAssertions::Yes) const
     {
         initializeIfNeeded(object);
 
-        ASSERT(&object == m_impl->template get<T>());
-        return WeakPtr<U, Counter>(*m_impl, enableWeakPtrThreadingAssertions);
+        ASSERT(&object == m_impl.pointer()->template get<T>());
+        return WeakPtr<U, Counter>(*m_impl.pointer(), enableWeakPtrThreadingAssertions);
     }
 
     void revokeAll()
     {
-        if (!m_impl)
-            return;
-
-        m_impl->clear();
-        m_impl = nullptr;
+        if (auto* pointer = m_impl.pointer()) {
+            pointer->clear();
+            m_impl.setPointer(nullptr);
+        }
     }
 
-    unsigned weakPtrCount() const { return m_impl ? m_impl->refCount() - 1 : 0; }
+    unsigned weakPtrCount() const
+    {
+        if (auto* pointer = m_impl.pointer())
+            return pointer->refCount() - 1;
+        return 0;
+    }
 
 #if ASSERT_ENABLED
-    bool isInitialized() const { return m_impl; }
+    bool isInitialized() const { return m_impl.pointer(); }
 #endif
 
+    uint16_t bitfield() const { return m_impl.type(); }
+    void setBitfield(uint16_t value) const { return m_impl.setType(value); }
+
 private:
     template<typename, typename, EnableWeakPtrThreadingAssertions> friend class WeakHashSet;
     template<typename, typename, typename> friend class WeakHashMap;
     template<typename, typename> friend class WeakPtr;
 
-    mutable RefPtr<WeakPtrImpl<Counter>> m_impl;
+    mutable CompactRefPtrTuple<WeakPtrImpl<CounterType>, uint16_t> m_impl;
```

Comment:
The way I was thinking of this, we can specify WeakPtrImpl type the same way we specify T.
So, WeakPtr usage becomes WeakPtr<T, WeakPtrImpl = DefaultWeakPtrImpl>.
Then WeakPtrImpl can be specialized for EventTarget.

Reply:
Sounds good.